edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c

```diff
@@ -27,170 +32,105 @@ typedef struct CERT_DATA_INFO_TAG
-            if (begin_hdr_end)
+            // check for first cert
+            if (!first_cert_begin_marker_found)
             {
-                // Once we are in the header then, copy the cert excluding \r\n on all lines
-                if (*iterator != '\r' && *iterator != '\n')
+                if ((ptr = strstr(iterator, BEGIN_CERT_MARKER)) != NULL)
                 {
-                    cert_base64[cert_idx++] = *iterator;
+                    first_start = ptr;
+                    first_cert_begin_marker_found = true;
+                    iterator = ptr + BEGIN_CERT_MARKER_LEN - 1;
                 }
-                if (*iterator == '\n' && *(iterator + 1) == '-')
+            }
+            else
+            {
+                if ((ptr = strstr(iterator, END_CERT_MARKER)) != NULL)
                 {
-                    // mark the end of the first certificate including \r\n characters
-                    cert_info->first_cert_end = iterator + END_HEADER_LENGTH + 1;
-                    if (*(cert_info->first_cert_end) == '\r')
+                    first_end = ptr + END_CERT_MARKER_LEN - 1;
+                    if ((first_end + 1 <= pem_end) && (*(first_end + 1) == '\r'))
                     {
-                        cert_info->first_cert_end++;
+                        first_end++;
                     }
-
-                    // Check to see if we have a chain embedded in the certificate
-                    // if we've have more data after the END HEADER then we have a chain
-                    if ((((iterator - cert_info->certificate_pem) + END_HEADER_LENGTH) + begin_hdr_len) < (int)len)
+                    if ((first_end + 1 <= pem_end) && (*(first_end + 1) == '\n'))
                     {
-                        iterator++;
-                        // Find the certificate chain here for later use
-                        while (*iterator != '\0')
-                        {
-                            // check for end header
-                            if (*iterator == '\n')
-                            {
-                                cert_info->cert_chain = iterator + 1;
-                                break;
-                            }
-                            iterator++;
-                        }
+                        first_end++;
                     }
-
-                    // If we encounter the \n- to signal the end header break out
-                    break;
+                    first_cert_found = true;
+                    iterator = first_end;
                 }
             }
-            else if (!begin_hdr_end && *iterator == '\n')
-            {
-                // Loop through the cert until we get to the \n at the end
-                // of the header
-                begin_hdr_end = true;
-            }
-            else
+        }
+        else
+        {
+            // now find the start and end of the chain
+            if ((ptr = strstr(iterator, BEGIN_CERT_MARKER)) != NULL)
```

Comment:
If this `strstr` returns `NULL` here, should we continue to advance iterator and search, or should we just `break`? I think if we don't find what we're looking for the first time we call this, future calls won't find it either.

Reply:
good point, I will add the fix.

Reply:
fixed.